src/cbor_decode.c

```diff
@@ -0,0 +1,390 @@
+				size_t *const p_elem_count,
+				uint8_t *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	uint8_t major_type = **pp_payload >> 5;
+	uint32_t val;
+
+
+	if (major_type != CBOR_MAJOR_TYPE_PRIM) {
+		/* Value to be read doesn't have the right type. */
+		cbor_decode_trace();
+		return false;
+	}
+	if (!uint32_decode(pp_payload, p_payload_end, p_elem_count,
+			&val, p_min_result, p_max_result)) {
+		cbor_decode_trace();
+		return false;
+	}
+	if (p_result != NULL) {
+		*p_result = val;
+	}
+	return true;
+}
+
+
+bool boolx_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				bool *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	uint8_t min_result = *(uint8_t *)p_min_result + 20;
+	uint8_t max_result = *(uint8_t *)p_max_result + 20;
+
+	if (!primx_decode(pp_payload, p_payload_end, p_elem_count,
+			(uint8_t *)p_result, &min_result, &max_result)) {
+		cbor_decode_trace();
+		return false;
+	}
+	(*p_result) -= 20;
+	return true;
+}
+
+
+bool double_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				double *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	uint8_t major_type = **pp_payload >> 5;
+
+	if (major_type != CBOR_MAJOR_TYPE_PRIM) {
+		/* Value to be read doesn't have the right type. */
+		cbor_decode_trace();
+		return false;
+	}
+	if (!value_extract(pp_payload, p_payload_end, p_elem_count, p_result,
+			sizeof(*p_result))) {
+		cbor_decode_trace();
+		return false;
+	}
+
+	if (!COMP_PTRS(double, p_result, p_min_result, p_max_result)) {
+		cbor_decode_trace();
+		return false;
+	}
+	return true;
+}
+
+
+bool any_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				void *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	if (p_result != NULL) {
+		/* 'any' type cannot be returned, only skipped. */
+		cbor_decode_trace();
+		return false;
+	}
+
+	uint8_t major_type = **pp_payload >> 5;
+	uint64_t value;
+
+	if (!value_extract(pp_payload, p_payload_end, p_elem_count, &value,
+			8)) {
+		/* Should not happen? */
+		cbor_decode_trace();
+		return false;
+	}
+
+	switch (major_type) {
+		case CBOR_MAJOR_TYPE_BSTR:
+		case CBOR_MAJOR_TYPE_TSTR:
+		case CBOR_MAJOR_TYPE_LIST:
+		case CBOR_MAJOR_TYPE_MAP:
+			(*pp_payload) += value;
+			break;
+		default:
+			/* Do nothing */
+			break;
+	}
+
+	return true;
+}
+
+
+bool multi_decode(size_t min_decode, size_t max_decode,
+				size_t *p_num_decode, decoder_t decoder,
+				uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				void *p_result, void *p_min_result,
+				void *p_max_result, size_t result_len)
+{
+	for (size_t i = 0; i < max_decode; i++) {
+		uint8_t const *p_payload_bak = *pp_payload;
+		size_t elem_count_bak = *p_elem_count;
+
+		if (!decoder(pp_payload, p_payload_end, p_elem_count,
```

Comment:
Single line per parameter? (I prefer being able to equate of the inner decode parameters to the outer function parameters)

Reply:
Done